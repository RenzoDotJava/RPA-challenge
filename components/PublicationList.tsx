import { useState, useCallback } from "react"
import { StyleSheet, Platform, StatusBar, SectionList, RefreshControl } from 'react-native'
import { UserPublications } from "../interfaces/PublicationInterfaces";
import { getUsersPublications } from "../services/publicationService";
import PublicationContent from './PublicationContent';
import PublicationTitle from './PublicationTitle';

interface PublicationListProps {
  filteredUsersPublications: UserPublications[];
  setUsersPublications: (list: UserPublications[]) => void;
  setFilteredUsersPublications: (list: UserPublications[]) => void;
  setSearch: (text: string) => void;
}

const PublicationList = ({filteredUsersPublications, setUsersPublications, setFilteredUsersPublications, setSearch}: PublicationListProps) => {
  const [loading, setLoading] = useState<boolean>(false);

  const onRefresh = useCallback(() => {
		setLoading(true);
		getUsersPublications(true)
			.then((res: UserPublications[]) => {
				setUsersPublications(res);
				setFilteredUsersPublications(res);
			})
			.catch(() => {
				setUsersPublications([]);
				setFilteredUsersPublications([]);
			})
			.finally(() => {
				setSearch("");
				setLoading(false);
			});
	}, []);

  return (
    <SectionList
      refreshControl={
        <RefreshControl
          refreshing={loading}
          onRefresh={onRefresh}
          colors={["#32373A"]}
          tintColor={"#32373A"}
        />
      }
      sections={filteredUsersPublications}
      keyExtractor={(item) => item.id.toString()}
      renderItem={({ item, index }) => (
        <PublicationContent
          index={index}
          title={item.title}
          body={item.body}
        />
      )}
      renderSectionHeader={({ section: { userId } }) => (
        <PublicationTitle userId={userId} />
      )}
      stickySectionHeadersEnabled={false}
    />
  )
}

export default PublicationList