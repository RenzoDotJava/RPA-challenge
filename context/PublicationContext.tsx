import { createContext } from "react";
import { IUserPublications } from "../interfaces/publication.interface";

type PublicationContextProps = {
	filteredUsersPublications: IUserPublications[];
	search: string;
	loading: boolean;
	getUsersPublications: (refresh: boolean) => Promise<IUserPublications[]>;
	searchFilter: (text: string) => void;
	onRefreshPublications: () => void;
};

export const PublicationContext = createContext<PublicationContextProps>(
	{} as PublicationContextProps
);
