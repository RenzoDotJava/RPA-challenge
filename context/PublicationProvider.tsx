import AsyncStorage from "@react-native-async-storage/async-storage";
import { useState, useEffect, useCallback } from "react";
import { IPublication, IUserPublications } from "../interfaces/publication.interface";
import { PublicationContext } from "./PublicationContext";
import { PublicationService } from "../services/publicationService";

export const PublicationProvider: React.FC<any> = ({ children }) => {
	const [usersPublications, setUsersPublications] = useState<IUserPublications[]>([]);
	const [filteredUsersPublications, setFilteredUsersPublications] = useState<IUserPublications[]>([]);
  const [search, setSearch] = useState<string>("");
  const [loading, setLoading] = useState<boolean>(false);

  const publicationService: PublicationService = new PublicationService();

  useEffect(() => {
		let isCanceled: boolean = false;

		getUsersPublications().then((res: IUserPublications[]) => {
			if (isCanceled) return;

			setUsersPublications(res);
			setFilteredUsersPublications(res);
		});

		return () => {
			isCanceled = true;
		};
	}, []);

  const groupPublicationsByUserId = (
    publications: IPublication[]
  ): IUserPublications[] => {
    return publications.reduce((acc: IUserPublications[], val: IPublication) => {
      let publication: IPublication = {
        id: val.id,
        title: val.title,
        body: val.body.replace(/\n|\r/g, ""),
      };
  
      const existingUser: IUserPublications | undefined = acc.find(
        (group: IUserPublications) => group.userId === val.userId
      );
  
      if (existingUser) {
        existingUser.data.push(publication);
      } else {
        acc.push({
          userId: val.userId,
          data: [publication],
        });
      }
  
      return acc;
    }, []);
  };

  const getPublicationsByUsersAmount = (
    amount: number = 5
  ): Promise<IUserPublications[]> => {
    let indexes: number[] = Array.from({ length: amount }, (val, i) => i);
    return Promise.all(
      indexes.map((index) => {
        return publicationService.getPublicationsByUserId(index + 1);
      })
    ).then(async (res: IPublication[][]) => {
      const groupedPublications: IUserPublications[] =
        groupPublicationsByUserId(res.flat());
  
      await AsyncStorage.setItem(
        "usersPublications",
        JSON.stringify(groupedPublications)
      );
  
      return groupedPublications;
    });
  };

  const getUsersPublications = async (
    refresh: boolean = false
  ): Promise<IUserPublications[]> => {
    let json: string | null = "";
    let usersPublications: IUserPublications[] = [];
  
    if (!refresh) json = await AsyncStorage.getItem("usersPublications");
  
    if (!json || json == "") {
      await getPublicationsByUsersAmount().then((res) => {
        usersPublications = res;
      });
    } else {
      usersPublications = JSON.parse(json);
    }
  
    return usersPublications;
  };

  const searchFilter = (text: string) => {
		if (text.trim() == "") {
			setFilteredUsersPublications(usersPublications);
		} else {
			const newPublications: IUserPublications[] = usersPublications
				.filter((item) =>
					item.data.some((d) => d.title.includes(text.toLowerCase()))
				)
				.map((item) => ({
					userId: item.userId,
					data: item.data.filter((d) =>
						d.title.includes(text.toLowerCase())
					),
				}));

			setFilteredUsersPublications(newPublications);
		}

		setSearch(text);
	};

  const onRefreshPublications = useCallback(() => {
		setLoading(true);
		getUsersPublications(true)
			.then((res: IUserPublications[]) => {
				setUsersPublications(res);
				setFilteredUsersPublications(res);
			})
			.catch(() => {
				setUsersPublications([]);
				setFilteredUsersPublications([]);
			})
			.finally(() => {
				setSearch("");
				setLoading(false);
			});
	}, []);

	return (
		<PublicationContext.Provider
			value={{
				filteredUsersPublications,
        search,
        loading,
        getUsersPublications,
        searchFilter,
        onRefreshPublications
			}}
		>
			{children}
		</PublicationContext.Provider>
	);
};
