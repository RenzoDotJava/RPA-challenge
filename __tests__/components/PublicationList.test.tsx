import { render, act, waitFor, screen } from "@testing-library/react-native";
import PublicationList from "../../components/PublicationList";
import { PublicationContext } from "../../context/PublicationContext";

const userPublications = [
	{
		userId: 1,
		data: [
			{
				id: 1,
				title: "title mockup 1",
				body: "body mockup 1",
			},
		],
	},
	{
		userId: 2,
		data: [
			{
				id: 2,
				title: "title mockup 2",
				body: "body mockup 2",
			},
		],
	},
];

describe("Testing <PublicationList />", () => {
	beforeEach(() => jest.clearAllMocks());

	it("Matching snapshot", () => {
		render(
			<PublicationContext.Provider
				value={{
					filteredUsersPublications: [],
					search: "",
					loading: false,
					getUsersPublications: jest.fn(),
					searchFilter: jest.fn(),
					onRefreshPublications: jest.fn()
				}}
			>
				<PublicationList />
			</PublicationContext.Provider>
		);

		act(() => {
			expect(screen).toMatchSnapshot();
		});

		waitFor(() => {
			expect(screen.getByTestId("publication-list")).toBeDefined();
		});
	});

	it("Showing the empty list", () => {
		render(
			<PublicationContext.Provider
				value={{
					filteredUsersPublications: [],
					search: "",
					loading: false,
					getUsersPublications: jest.fn(),
					searchFilter: jest.fn(),
					onRefreshPublications: jest.fn()
				}}
			>
				<PublicationList />
			</PublicationContext.Provider>
		);

		expect(screen.getByText("No hay publicaciones")).toBeTruthy();
	});

	it("Showing the publication list", () => {
		render(
			<PublicationContext.Provider
				value={{
					filteredUsersPublications: userPublications,
					search: "",
					loading: false,
					getUsersPublications: jest.fn(),
					searchFilter: jest.fn(),
					onRefreshPublications: jest.fn()
				}}
			>
				<PublicationList />
			</PublicationContext.Provider>
		);

		expect(screen.queryAllByTestId("publication-title").length).toBe(
			userPublications.length
		);
	});

	it("Running onRefresh function when doing pull to refresh", async () => {

    const onRefreshMockup = jest.fn();

		render(
			<PublicationContext.Provider
				value={{
					filteredUsersPublications: userPublications,
					search: "",
					loading: false,
					getUsersPublications: jest.fn(),
					searchFilter: jest.fn(),
					onRefreshPublications: onRefreshMockup
				}}
			>
				<PublicationList />
			</PublicationContext.Provider>
		);

		const publicationList = screen.getByTestId("publication-list");
		expect(publicationList).toBeDefined();

		const { refreshControl } = publicationList.props;

		await act(async () => {
			refreshControl.props.onRefresh();
		});

		expect(onRefreshMockup).toHaveBeenCalled();
	});
});
