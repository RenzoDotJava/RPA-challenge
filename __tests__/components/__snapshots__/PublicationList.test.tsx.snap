// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <PublicationList /> Matching snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#CCCCCC",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={
      <EmptyList
        text="No hay publicaciones"
      />
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        colors={
          [
            "#32373A",
          ]
        }
        onRefresh={[MockFunction]}
        refreshing={false}
        tintColor="#32373A"
      />
    }
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="publication-list"
  >
    <RCTRefreshControl />
    <View>
      <View
        style={
          {
            "flex": 1,
            "marginTop": 15,
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Arial",
              "fontSize": 20,
              "fontWeight": "500",
              "textAlign": "center",
            }
          }
        >
          No hay publicaciones
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
